crates/mod-aws-sigv4/src/lib.rs

```diff
@@ -0,0 +1,315 @@
+        .collect();
+    encoded_params.sort();
+
+    encoded_params
+        .iter()
+        .map(|(k, v)| format!("{}={}", k, v))
+        .collect::<Vec<_>>()
+        .join("&")
+}
+
+fn create_canonical_headers(headers: &BTreeMap<String, String>) -> (String, String) {
+    // Convert headers to lowercase and trim values
+    let canonical_headers: BTreeMap<String, String> = headers
+        .iter()
+        .map(|(k, v)| (k.to_lowercase(), v.trim().to_string()))
+        .collect();
+
+    // Sort headers
+    let header_string = canonical_headers
+        .iter()
+        .map(|(k, v)| format!("{}:{}", k, v))
+        .collect::<Vec<_>>()
+        .join("\n");
+
+    // Create signed headers list
+    let signed_headers = canonical_headers
+        .keys()
+        .cloned()
+        .collect::<Vec<_>>()
+        .join(";");
+
+    (header_string, signed_headers)
+}
+
+fn create_signing_key(secret_key: &str, date_stamp: &str, region: &str, service: &str) -> Vec<u8> {
+    let k_date = hmac_sha256(
+        format!("AWS4{}", secret_key).as_bytes(),
+        date_stamp.as_bytes(),
+    );
+    let k_region = hmac_sha256(&k_date, region.as_bytes());
+    let k_service = hmac_sha256(&k_region, service.as_bytes());
+    hmac_sha256(&k_service, b"aws4_request")
+}
+
+async fn sign_request(req: SigV4Request) -> anyhow::Result<SigV4Response> {
+    // Get the secret key
+    let secret_key_bytes = req.secret_key.get().await?;
+    let secret_key = std::str::from_utf8(&secret_key_bytes)
+        .context("secret_key must be valid UTF-8")?
+        .to_string();
+
+    // Use provided timestamp or current time
+    let timestamp = req.timestamp.unwrap_or_else(Utc::now);
+    let amz_date = timestamp.format("%Y%m%dT%H%M%SZ").to_string();
+    let date_stamp = timestamp.format("%Y%m%d").to_string();
+
+    // Create payload hash
+    let payload_hash = sha256_hex(req.payload.as_bytes());
+
+    // Prepare headers - add required AWS headers
+    let mut headers = req.headers.clone();
+    headers.insert("host".to_string(), "".to_string()); // Will be set by caller
+    headers.insert("x-amz-date".to_string(), amz_date.clone());
+
+    if let Some(token) = &req.session_token {
+        headers.insert("x-amz-security-token".to_string(), token.clone());
+    }
+
+    // Add content hash header for some services
+    if req.service != "s3" {
+        headers.insert("x-amz-content-sha256".to_string(), payload_hash.clone());
+    }
+
+    // Create canonical request
+    let canonical_uri = create_canonical_uri(&req.uri);
+    let canonical_query_string = create_canonical_query_string(&req.query_params);
+    let (canonical_headers, signed_headers) = create_canonical_headers(&headers);
+
+    let canonical_request = format!(
+        "{}\n{}\n{}\n{}\n\n{}\n{}",
+        req.method,
+        canonical_uri,
+        canonical_query_string,
+        canonical_headers,
+        signed_headers,
+        payload_hash
+    );
```

Comment:
This form is a bit more readable, especially since there are so many parameters!

```suggestion
    let canonical_request = format!(
        "{method}\n{canonical_uri}\n{canonical_query_string}\n{canonical_headers}\n\n{signed_headers}\n{payload_hash}",
        method = req.method,
    );
```